lightning/src/ln/chan_utils.rs

```diff
@@ -60,18 +60,56 @@ pub fn htlc_timeout_tx_weight(opt_anchors: bool) -> u64 {
 }
 
 #[derive(PartialEq)]
-pub(crate) enum HTLCType {
-	AcceptedHTLC,
-	OfferedHTLC
+pub(crate) enum HTLCClaim {
+	OfferedTimeout,
+	OfferedPreimage,
+	AcceptedTimeout,
+	AcceptedPreimage,
+	Revocation,
 }
 
-impl HTLCType {
-	/// Check if a given tx witnessScript len matchs one of a pre-signed HTLC
-	pub(crate) fn scriptlen_to_htlctype(witness_script_len: usize) ->  Option<HTLCType> {
-		if witness_script_len == 133 {
-			Some(HTLCType::OfferedHTLC)
-		} else if witness_script_len >= 136 && witness_script_len <= 139 {
-			Some(HTLCType::AcceptedHTLC)
+impl HTLCClaim {
+	/// Check if a given input witness attempts to claim a HTLC.
+	pub(crate) fn from_witness(witness: &Witness) -> Option<Self> {
+		if witness.len() < 2 {
+			return None;
+		}
+		let witness_script = witness.last().unwrap();
+		let second_to_last = witness.second_to_last().unwrap();
+		if witness_script.len() == 133 {
+			if witness.len() == 3 && second_to_last.len() == 33 {
+				Some(Self::Revocation)
+			} else if witness.len() == 3 && second_to_last.len() == 32 {
+				Some(Self::OfferedPreimage)
+			} else if witness.len() == 5 && second_to_last.len() == 0 {
+				Some(Self::OfferedTimeout)
+			} else {
+				None
+			}
+		} else if witness_script.len() == 136 {
+			if witness.len() == 3 && second_to_last.len() == 33 {
+				Some(Self::Revocation)
+			} else if witness.len() == 3 && second_to_last.len() == 32 {
+				Some(Self::OfferedPreimage)
+			} else if witness.len() == 5 && second_to_last.len() == 0 {
+				Some(Self::OfferedTimeout)
+			} else if witness.len() == 3 && second_to_last.len() == 0 {
+				Some(Self::AcceptedTimeout)
+			} else if witness.len() == 5 && second_to_last.len() == 32 {
+				Some(Self::AcceptedPreimage)
+			} else {
+				None
+			}
+		} else if witness_script.len() >= 137 && witness_script.len() <= 143 {
```

Comment:
I think here it could be valuable to a) constify the value, actually we already do it in `functional_test_utils.rs` for those exact values and b) for each branch document the type of outputs and witness stack to ensure detection is actually correct. #1585 issue was affecting this exact code path.